aproxy.go

```diff
@@ -0,0 +1,404 @@
+//go:build linux
+
+package main
+
+import (
+	"bufio"
+	"encoding/binary"
+	"errors"
+	"flag"
+	"fmt"
+	"golang.org/x/crypto/cryptobyte"
+	"io"
+	"log"
+	"log/slog"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"syscall"
+	"unsafe"
+)
+
+var version = "0.2.0"
+
+type PrereadConn struct {
+	prereadStarted bool
+	prereadEnded   bool
+	prereadBuf     []byte
+	mu             sync.Mutex
+	conn           net.Conn
+}
+
+func (c *PrereadConn) StartPreread() {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if c.prereadStarted {
+		panic("call StartPreread after preread has already started or ended")
+	}
+	c.prereadStarted = true
+}
+
+func (c *PrereadConn) RestorePreread() {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if !c.prereadStarted || c.prereadEnded {
+		panic("call RestorePreread after preread has ended or hasn't started")
+	}
+	c.prereadEnded = true
+}
+
+func (c *PrereadConn) Read(p []byte) (n int, err error) {
```

Comment:
A couple of comments in this function explaining what it does would be helpful. Like "first reads blah, then once the preread is started, does blah" and so on.

Reply:
Added, thanks!